massifs/localmassifreader.go

```diff
@@ -0,0 +1,384 @@
+		MMRState:     vc.MMRState,
+	})
+}
+
+// GetMassif reads the massif identified by the tenant identity and massif index
+func (r *LocalReader) GetMassif(
+	ctx context.Context, tenantIdentityOrLocalPath string, massifIndex uint64,
+	opts ...ReaderOption,
+) (MassifContext, error) {
+
+	// short circuit direct match, regardless of mode, to support explicit paths
+	dirEntry, ok := r.cache.GetEntry(tenantIdentityOrLocalPath)
+	if ok {
+		return copyCachedMassifOrErr(dirEntry.ReadMassif(r.cache, massifIndex))
+	}
+
+	directory, err := r.ResolveMassifDir(tenantIdentityOrLocalPath)
+	if err != nil {
+		return MassifContext{}, err
+	}
+
+	return copyCachedMassifOrErr(r.cache.ReadMassif(directory, massifIndex))
+}
+
+// GetSeal reads the seal identified by the tenant identity and massif index
+func (r *LocalReader) GetSeal(
+	ctx context.Context, tenantIdentityOrLocalPath string, massifIndex uint64,
+	opts ...ReaderOption,
+) (SealedState, error) {
+
+	// short circuit direct match, regardless of mode, to support explicit paths
+	dirEntry, ok := r.cache.GetEntry(tenantIdentityOrLocalPath)
+	if ok {
+		return copyCachedSealOrErr(dirEntry.ReadSeal(r.cache, tenantIdentityOrLocalPath))
+	}
+
+	directory, err := r.ResolveSealDir(tenantIdentityOrLocalPath)
+	if err != nil {
+		return SealedState{}, err
+	}
+
+	return copyCachedSealOrErr(r.cache.ReadSeal(directory, massifIndex))
+}
+
+// GetSignedRoot satisfies the SealGetter interface.
+// This is the default seal getter for the local reader when GetVerifiedContext is called.
+func (r *LocalReader) GetSignedRoot(
+	ctx context.Context, tenantIdentityOrLocalPath string, massifIndex uint32,
+	opts ...ReaderOption,
+) (*cose.CoseSign1Message, MMRState, error) {
+	sealedState, err := r.GetSeal(ctx, tenantIdentityOrLocalPath, uint64(massifIndex), opts...)
+	if err != nil {
+		return nil, MMRState{}, err
+	}
+	return &sealedState.Sign1Message, sealedState.MMRState, nil
+}
+
+// GetHeadMassif reads the most recent massif in the log identified by the tenant identity
+func (r *LocalReader) GetHeadMassif(
+	ctx context.Context, tenantIdentityOrLocalPath string,
+	opts ...ReaderOption,
+) (MassifContext, error) {
+
+	dirEntry, ok := r.cache.GetEntry(tenantIdentityOrLocalPath)
+	if ok {
+		return copyCachedMassifOrErr(dirEntry.ReadMassif(r.cache, uint64(dirEntry.HeadMassifIndex)))
+	}
+
+	directory, err := r.cache.ResolveMassifDir(tenantIdentityOrLocalPath)
+	if err != nil {
+		return MassifContext{}, err
+	}
+	err = r.cache.FindMassifFiles(directory)
+	if err != nil {
+		return MassifContext{}, err
+	}
+	dirEntry, ok = r.cache.GetEntry(directory)
+	if !ok {
+		// It's a bug in FindMassifFiles if this case actually happens in practice.
+		return MassifContext{}, fmt.Errorf("failed to prime directory: %s", directory)
+	}
+	return copyCachedMassifOrErr(dirEntry.ReadMassif(r.cache, uint64(dirEntry.HeadMassifIndex)))
+}
+
+func (r *LocalReader) GetFirstMassif(
+	ctx context.Context, tenantIdentityOrLocalPath string,
```

Comment:
Yes, that is the essence of the interface change. We change the meaning of that parameter to be "the identifier needed to locate the massif data". For the regular massif interface that is just the tenant identity. For 

I did consider completely changing the interface so that we just require "path" everywhere, but that seemed to far reaching, and I'm actually fine with the interpretation of the identifier being reader implementation specific.

I can't see any way out of this if we want readers with the same interface to support blobs, local files on disc, and potentially a local sqlite db or something

Reply:
yeh i agree, this is the best solution i can think of!